ostd/src/arch/x86/allocator.rs

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{borrow::ToOwned, vec::Vec};
+
+use align_ext::AlignExt;
+
+use crate::{boot::memory_region::MemoryRegionType, io::io_mem::allocator::IoMemAllocatorBuilder};
+
+/// Initializes the allocatable MMIO area based on the x86-64 memory distribution map.
+///
+/// In x86-64, the available physical memory area is divided into two regions below 32 bits (Low memory)
+/// and above (High memory). The area from the top of low memory to 0xffff_ffff and the area after the
+/// top of high memory are available MMIO areas.
+pub(super) fn construct_io_mem_allocator_builder() -> IoMemAllocatorBuilder {
+    // TODO: Add MMIO regions below 1MB (e.g., VGA framebuffer).
+    let reversed_regions = {
+        let mut regions = crate::boot::memory_regions().to_owned();
+        assert!(!regions.is_empty());
+        regions.sort_by_key(|a| a.base());
```

Comment:
This `construct_io_mem_allocator_builder` function should be refactored for a simpler and more elegant implementation. Sorting all the regions has a complexity of $O(N\log{N})$, but finding the highest regions under 32-bit and 64-bit address spaces only need two scans of $O(N)$ complexity. So I expect an iterator-based, concise implementation.

```rust
let all_regions = crate::boot::memory_regions();

let highest_region_in_lower_half = all_regions
    .iter()
    .filter(|r| r.base() < u32::MAX as usize)
    .max_by(|a, b| a.base().cmp(b.base()));
// ...

let highest_region_in_upper_half = all_regions
    .iter()
    .filter(|r| r.base() >= u32::MAX as usize)
    .max_by(|a, b| a.base().cmp(b.base()));
// ...
```